specification/dsdl/dsdl.tex

```diff
@@ -132,30 +132,100 @@ \section{Syntax}\label{sec:dsdl_syntax}
 directives apply only to the part of the service type definition where they are defined,
 not crossing the boundary of the service response marker.
 
-\subsection{Attribute definition}
+\subsection{Keywords}\label{sec:dsdl_keywords}
+A comprehensive list of non context sensitive keywords and a short description is given below, 
+a bold face \texttt{\textbf{X}} is interchangeable with a decimal number of one or more digits (\texttt{[0-9]+}).
+
+\begin{UAVCANSimpleTable}{A comprehensive table of DSDL keywords}{|l l|}
+\textbf{Keyword} & \textbf{Description} \\
+\texttt{saturated} & Cast mode. \\
+\texttt{truncated} & Cast mode. \\
+\texttt{offset} & Usable only in assertions, evaluates into the set of possible bit offsets. \\
+\texttt{int\textbf{X}} & Primitive type. \\
+\texttt{uint\textbf{X}} & Primitive type. \\
+\texttt{void\textbf{X}} & Primitive type. \\
+\texttt{bool} & Primitive type. \\
+\texttt{true} & \texttt{bool} variant. \\
+\texttt{false} & \texttt{bool} variant. \\
+\texttt{con} & Reserved. \\
+\texttt{prn} & Reserved. \\
+\texttt{aux} & Reserved. \\
+\texttt{nul} & Reserved. \\
+\texttt{com\textbf{X}} & Reserved. \\
+\texttt{lpt\textbf{X}} & Reserved. \\
+\end{UAVCANSimpleTable}
```

Comment:
Yes, let's reserve these. I agree.

> The word offset should be un-reserved because it is too useful on its own (e.g. see the file transfer types). As it is only useful in expressions, it is perfectly safe to use this word for naming attributes and types.

Then we would have a contextual keyword, which makes implementation unnecessary hard/messy. I don't think this is a good idea, let's instead use a different word or a symbol for this then?

Reply:
I also added `enum`

Reply:
> Then we would have a contextual keyword, which makes implementation unnecessary hard/messy.

Eh, I don't think it does. `offset` doesn't need to be a keyword, it is an identifier which refers to a particular value. Like a field. It doesn't require any special handling logic, and hence it doesn't need to be reserved. Replacing `offset` with a special character or prefixing it with special characters is probably not a good idea because readability will suffer.

Reply:
> If offsets are replaced with constant set literals early on, it needs to be figured out whether we're in an assertion context before they can be replaced or not.

I don't think we gain much by allowing early replacement. Our grammar remains simple enough even if we postpone offset handling until the point where assertions are checked. Assertion checking must be postponed until the very end of the parsing process anyway, because until all of the fields are processed, their sizes and offsets cannot be known.

Parsers can treat `@assert` (and `@print`) directives as black boxes until everything else is parsed, not even looking at their contents until it can't be delayed anymore. This removes the need for special handling of any keywords inside. Perhaps you have a different architecture in mind?

Reply:
In the name of keeping DSDL simple, we should use a keyword that is fine to reserve. Things will probably be fine with context dependent keywords, but it's not a place where one normally wants to start.

The worst thing I can think off in 30 seconds is the following ambigouty:
```
uint8 offset = 12
uint8 field
@assert offset == {8}
```

We have to do one of the following:
 - Disallow the situation by making offset a keyword
 - Define that these "magic words" take precedence over consts.

This was pretty much all i had. If you insist on removing `offset` as a keyword, just say so and I will do it.

Reply:
Ah, you have a point. When I originally proposed assertion checks, I intended them to be much simpler than what they are becoming now; particularly, I did not plan for the scenario you described to be possible. I realize though that we later discussed it further and agreed that a more advanced expression evaluation capability should be implemented, so my original outlook is no longer valid.

Okay, can we somehow keep the keyword sensible and at the same time avoid robbing definitions from useful words like `offset`? Ugly things come to mind like `_offset`. The screaming snake case does not seem to help because there can always be a constant named `OFFSET`, we don't want to prohibit that either. Perhaps we could pick a less useful keyword and use it as an umbrella identifier for referring to special entities constructed by the parser; e.g. instead of `offset` we would say `meta.offset`? Or just `.offset`, although it might be a bit opaque.

Reply:
> Perhaps we could pick a less useful keyword and use it as an umbrella identifier for referring to special entities constructed by the parser; e.g. instead of offset we would say meta.offset? Or just .offset, although it might be a bit opaque.

What about introducing a concept of "meta values", and use a specific symbol for them (similar to compiler directives like `@assert`).  We can then call `offset` a location dependent meta value. We can later introduce things as `MAJOR_VERSION`, `PI`, etc if we want to.

What about some sort of bracket notation `<offset>`? Or a dollar symbol `$offset`?

Reply:
So we are choosing between `meta.offset` and `$offset`, if we were to ignore the ugly options like `_offset` or `<offset>`. I generally dislike overly short constructs because they might undermine clarity, but in this case choosing dollars over an umbrella identifier I suggested earlier might be sensible because it works well in other contexts: the dollar allows us to un-reserve several keywords that are used as special-purpose identifiers: `root`, `super`, `self`, and use `$root`, `$super`, and `$self` instead.

So the list of additional reserved keywords will include: `enum`, `struct`, `const`, `template`, and the current offset will be referred to as `$offset`. Any objections?

Reply:
Great!

I would like to keep `root`, `super` and `self` for now. If only for being super likely to collide with the target language. What do you think?

Reply:
I'm so used to '$' meaning "variable" (perl, bash, and Make all use this syntax) that it would confuse the crap out of me to see it used as "meta keyword". I'd vote for one of the following:
```
.offset      # dot syntax as if this is a member of the dsdl global namespace and we are
             # descending one level of variable scope to access it.

::offset     # Similar to dot but using C++-like namespace syntax

.            # I proposed this before; to use ld syntax for "the currently computed memory location"

dsdl.offset  # explicit access to a global "dsdl" namespace

dsdl::offset # same as above but with C++-like syntax.
```

Reply:
> I would like to keep `root`, `super` and `self` for now. If only for being super likely to collide with the target language. What do you think?

Okay, why `root` though? Is there any mainstream language where this is a reserved keyword? The other two seem fine.

> I'm so used to '$' meaning "variable" (perl, bash, and Make all use this syntax) that it would confuse the crap out of me to see it used as "meta keyword".

Okay, so we're back to `meta.offset`, except that instead of `meta` there's now `dsdl`, which is fine. We could also consider using `type`, which has the advantage of being a reserved keyword/identifier in Rust/Python: `type.offset`.

So, `dsdl` or `type`?

Reply:
I'm fine with either but I think I prefer dsdl.

Reply:
`dsdl.type` is ambiguous with an associated const from a different file (that doesn't adhere to naming conventions) called `dsdl` (in the same root namespace). 

I would prefer not to start reserving file names as well. It's just one more concept that doesn't need to exist.

Reply:
> This is (I guess?) unless we enforce all dsdl definitions to be version, I don't remember what we ended up on.

Yes, currently we require that whenever you refer to a type, you specify its version explicitly, always. But even that doesn't matter; my understanding is that reserved keywords can't be used for naming types or namespaces, otherwise conflicts and confusion are bound to occur. Even if the DSDL spec allowed you to get away with naming your type, say, `struct`, it is unlikely that you will be able to generate code out of that without fixing your identifiers first, like `struct_` or `r#struct` or whatever suits your language. If we accept this reasoning, then it is safe to use `type.offset`.

> Remember the `offset` symbol is not a const or a variable. It is some sort of magic context dependent that the compiler have special handling rules for.

We could rename it for clarity, like `type.current_offset`? I don't like re-using `@` because this symbol is not really a directive (the semantics are quite different), and `@assert_offset` is unnecessarily resctictive -- there are likely to be other things to check in the future, not just offsets; besides, usage of that construct in arithmetic expressions would be impossible or weird.

Reply:
> It is some sort of magic context dependent that the compiler have special handling rules for.

If we were to emphasize the magic nature of this value, and if $ is not an option, we could go back to the other options suggested by Scott above. Of which I don't like `.` (dot) because of its syntactic ambiguity (can't be parsed using regular grammars), so we're left with `::offset` or even just plain `:offset`. Any objections to that?

Reply:
I'm okay with either

Reply:
I'm also okay with either, small preference towards `:offset`